nifi-nar-bundles/nifi-hive-bundle/nifi-hive3-processors/src/main/java/org/apache/hive/streaming/NiFiRecordSerDe.java

```diff
@@ -142,7 +138,27 @@ public SerDeStats getSerDeStats() {
     public Object deserialize(Writable writable) throws SerDeException {
         ObjectWritable t = (ObjectWritable) writable;
         Record record = (Record) t.get();
-        List<Object> r = new ArrayList<>(Collections.nCopies(columnNames.size(), null));
+        return  deserialize(record, schema, true);
+    }
+
+    /**
+     * Deserialize a record object into a Hive struct.
+     * @param  record The record to deserialize, can be null
+     * @param structTypeInfo The hive table column info that corresponds to the Hive struct
+     * @param isParentStruct Whether or not this struct is the one contained by the Writable
+     */
+    Object deserialize(Record record, StructTypeInfo structTypeInfo, boolean isParentStruct) throws SerDeException{
+        if(record == null){
+            return  null;
+        }
+        Map<String, Integer> fieldPositionMap = null;
+        try {
+            fieldPositionMap = populateFieldPositionMap(record.getSchema(), structTypeInfo, log);
```

Comment:
This was previously done in `initialize()` because it only needed to be done once per flowfile, rather than doing the same work every time a record is deserialized. Is there a reason it needs to be moved into `deserialize()`?

Reply:
The problem was that `deserialize` could call itself if the field value was a nested record, array/list of records or a map of key->Record. This bothered me quite a bit but I didn't know how exactly to handle it.  My initial thoughts was to have a cache (a map basically) that I populate as I encounter new schema/struct type info; something to consider, we can have 2 exact schemas but different positions (imaging an account containing a list of accounts where the column names were re-ordered in the child collection) so the cache will need to handle that.

Reply:
Looks like I over-simplified the NiFiRecordSerDe code while porting from JsonSerDe. The latter had broken up the tasks of getting field names, extracting field values, and possibly recursively calling one or both of those methods for nested structures. I think we should look at [JsonSerDe](https://github.com/apache/hive/blob/rel/release-3.1.0/serde/src/java/org/apache/hadoop/hive/serde2/JsonSerDe.java) and do NiFiRecordSerDe likewise. It can result in getting that fieldPositionMap called multiple times per flow file, but like you said I think we need to do it. However they have a pretty good modular approach to which methods do which things, all we should have to do is map our Record methods to JsonSerDe's JSON parsing methods (`nextToken()` and such). Thoughts?

Reply:
I've checked out their code, will refactor to match what they've done